### PR TITLE
I/O Manage wrapped in an Arc<Mutex>

### PR DESCRIPTION
**Pull request by andreeaflorescu**
- Wrapped the I/O manager in an Arc<Mutex>
- Some other simplifications

Details in commit message.

### CODE REVIEW

**Comment by andreeaflorescu** [comment]
Hmm, this approach might still not work because we used the really simplified approach of adding devices by using the kernel command line.

File: src/vm-vcpu/src/vm.rs [hunk 1]
```diff
@@ -245,7 +249,9 @@ mod tests {
         // We're specifying in the VmState that we want to run one vcpu, but we do not create
         // any. This should return an error.
         let (mut vm, _guest_memory) = default_vm(1).unwrap();
-        assert!(matches!(vm.run(), Err(Error::RunVcpus(e)) if e.kind() == ErrorKind::InvalidInput));
+        assert!(
+            matches!(vm.run(GuestAddress(0x10000000)), Err(Error::RunVcpus(e)) if e.kind() == ErrorKind::InvalidInput)
```
Comment (aghecenco) [inline src/vm-vcpu/src/vm.rs:253]: nit (how did clippy miss this?): `0x1000_0000`

File: src/vmm/src/lib.rs [hunk 2]
```diff
@@ -386,25 +373,35 @@ impl VMM {
 mod tests {
     use super::*;
 
-    use std::io::ErrorKind;
     use std::path::PathBuf;
 
     use linux_loader::loader::bootparam::setup_header;
     use linux_loader::loader::elf::PvhBootCapability;
+    use linux_loader::loader::elf::PvhBootCapability::PvhEntryNotPresent;
```
Comment (aghecenco) [inline src/vmm/src/lib.rs:381]: nit: merge this `use` with the one above

File: src/vmm/src/lib.rs [hunk 3]
```diff
@@ -386,25 +373,35 @@ impl VMM {
...
+    use std::io::ErrorKind;
```
Comment (aghecenco) [inline src/vmm/src/lib.rs:382]: Ah, the old `use` reordering nit.

**Review by aghecenco — DISMISSED** [review]
LGTM, only style nits. One of these is a typo in 8596b8bc71c7b569d3ab53f9c666d934e30260af's commit message (s/`VMM::nwe`/`VMM::new`).

File: src/vm-vcpu/src/vm.rs (hunk 1 above)
Comment (andreeaflorescu) [reply src/vm-vcpu/src/vm.rs:253]: I don't think clippy runs on tests...

File: src/vmm/src/lib.rs (hunk 3 above)
Comment (andreeaflorescu) [reply src/vmm/src/lib.rs]: Done.

File: src/vmm/src/lib.rs (hunk 2 above)
Comment (andreeaflorescu) [reply src/vmm/src/lib.rs]: Done.

File: src/vm-vcpu/src/vm.rs (hunk 1 above)
Comment (andreeaflorescu) [reply src/vm-vcpu/src/vm.rs]: Done.

**Review by aghecenco — APPROVED** [review]

File: src/vmm/src/lib.rs [hunk 4]
```diff
@@ -333,17 +332,10 @@ impl VMM {
     /// # Arguments
     ///
     /// * `vcpu_cfg` - [`VcpuConfig`] struct containing vCPU configurations.
-    /// * `kernel_load` - [`KernelLoaderResult`] struct, result of loading the kernel in guest memory.
     ///
     /// [`KernelLoaderResult`]: https://docs.rs/linux-loader/latest/linux_loader/loader/struct.KernelLoaderResult.html
```
Comment (lauralt) [inline src/vmm/src/lib.rs:337]: nit: this should be removed as well.

File: src/vm-vcpu/src/vm.rs [hunk 5]
```diff
@@ -245,7 +249,9 @@ mod tests {
         // We're specifying in the VmState that we want to run one vcpu, but we do not create
         // any. This should return an error.
         let (mut vm, _guest_memory) = default_vm(1).unwrap();
-        assert!(matches!(vm.run(), Err(Error::RunVcpus(e)) if e.kind() == ErrorKind::InvalidInput));
+        assert!(
+            matches!(vm.run(GuestAddress(0x10_000_000)), Err(Error::RunVcpus(e)) if e.kind() == ErrorKind::InvalidInput)
```
Comment (lauralt) [inline src/vm-vcpu/src/vm.rs:253]: nit: I think clippy is suggesting to group the bytes like this: 0x1000_0000

File: src/vmm/src/lib.rs [hunk 6]
```diff
@@ -153,11 +154,11 @@ impl TryFrom<VMMConfig> for VMM {
             guest_memory,
             device_mgr: Arc::new(Mutex::new(IoManager::new())),
             event_mgr: EventManager::new().map_err(Error::EventManager)?,
+            kernel_cfg: config.kernel_config.clone(),
```
Comment (lauralt) [inline src/vmm/src/lib.rs:157]: The `clone` is not needed.

**Review by lauralt — APPROVED** [review]
Looks good, left a couple of nits.

File: src/vm-vcpu/src/vm.rs (hunk 5 above)
Comment (andreeaflorescu) [reply src/vm-vcpu/src/vm.rs:253]: Clippy doesn't seem to care a lot about the grouping, just that the group is not too large:
- https://rust-lang.github.io/rust-clippy/master/#large_digit_groups
- https://rust-lang.github.io/rust-clippy/master/#inconsistent_digit_grouping

**Pull request merged**